xc/domain.ml

```diff
@@ -1777,6 +1777,26 @@ let send_s3resume ~xc domid =
   debug "VM = %s; domid = %d; send_s3resume" (Uuid.to_string uuid) domid ;
   Xenctrlext.domain_send_s3resume xc domid
 
+let soft_reset ~xc ~xs domid =
+  let uuid = get_uuid ~xc domid in
+  debug "VM = %s; domid = %d; soft_reset" (Uuid.to_string uuid) domid;
+
+  pause ~xc domid;
+  Xenctrlext.domain_soft_reset xc domid;
+
+  let dom_path = xs.Xs.getdomainpath domid in
+  let store_mfn_s = xs.Xs.read (dom_path ^ "/store/ring-ref") in
+  let store_mfn = Nativeint.of_string store_mfn_s in
+  let store_port, console_port = create_channels ~xc uuid domid in
+  xs.Xs.introduce domid store_mfn store_port;
+
+  xs.Xs.write (dom_path ^ "/store/port") (string_of_int store_port);
```

Comment:
How about watches in Xenstore? They could fire at the wrong time while the guest is not yet prepared to handle it.
Would the guest be issuing the reset watch xenstored command? (if so, note that it is implemented incorrectly in oxenstored)

Reply:
Could you describe the scenario in more detail? I don't think guest would get any watch notification until it registers a handler again in kdump kernel.

Reply:
I enabled a new platform feature xs_reset_watches - that will tell the guest that it can reset watches for itself on kexec reboot. We just need to fix oxenstored ABI incompatibility and wait to test it.

Reply:
It is best effort operation in the guest - meaning it wouldn't break kexec boot if it's now attempted to be done by the guest. That means it doesn't block XAPI side enablement from merging.